internal/manifestwork/manifestwork.go

```diff
@@ -0,0 +1,120 @@
+package manifestwork
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	reflect "reflect"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	workv1 "open-cluster-management.io/api/work/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	kmmv1beta1 "github.com/kubernetes-sigs/kernel-module-management/api/v1beta1"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/constants"
+)
+
+//go:generate mockgen -source=manifestwork.go -package=manifestwork -destination=mock_manifestwork.go
+
+type ManifestWorkCreator interface {
+	GarbageCollect(ctx context.Context, clusters clusterv1.ManagedClusterList, mcm kmmv1beta1.ManagedClusterModule) error
+	SetManifestWorkAsDesired(ctx context.Context, mw *workv1.ManifestWork, mcm kmmv1beta1.ManagedClusterModule) error
+}
+
+type manifestWorkGenerator struct {
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+func NewCreator(client client.Client, scheme *runtime.Scheme) ManifestWorkCreator {
+	return &manifestWorkGenerator{
+		client: client,
+		scheme: scheme,
+	}
+}
+
+func (mwg *manifestWorkGenerator) GarbageCollect(ctx context.Context, clusters clusterv1.ManagedClusterList, mcm kmmv1beta1.ManagedClusterModule) error {
+	manifestWorks, err := mwg.getOwnedManifestWorks(ctx, mcm)
+	if err != nil {
+		return fmt.Errorf("failed to get owned ManifestWorks: %w", err)
+	}
+
+	clusterNames := map[string]string{}
+	for _, cluster := range clusters.Items {
+		clusterNames[cluster.Name] = ""
+	}
+
+	for _, mw := range manifestWorks.Items {
+		if _, contains := clusterNames[mw.Namespace]; contains {
+			continue
+		}
+
+		if err := mwg.client.Delete(ctx, &mw); err != nil {
+			return fmt.Errorf("failed to delete ManifestWork %s/%s: %w", mw.Namespace, mw.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func (mwg *manifestWorkGenerator) SetManifestWorkAsDesired(ctx context.Context, mw *workv1.ManifestWork, mcm kmmv1beta1.ManagedClusterModule) error {
+	if mw == nil {
+		return errors.New("mw cannot be nil")
+	}
+
+	// Ensure no Build and Sign specs reach the Spoke cluster
+	mcm.Spec.ModuleSpec.ModuleLoader.Container.Build = nil
+	mcm.Spec.ModuleSpec.ModuleLoader.Container.Sign = nil
```

Comment:
We should also go through all kernel mappings to strip those sections as well.

Reply:
Great catch, I completely missed this. I adjusted the code and also added the respective unit test case.